File: scripts/bloom-score.mjs
// Scores bloom variants on the two things that pull against each other: colour fidelity of the
// speedcube in a daylight close-up, and faithfulness to the established night look.
// Usage: VARIANTS='{"name":"query"}' node scripts/bloom-score.mjs <outDir>
import { chromium } from 'playwright-core';
const out = process.argv[2];
const variants = JSON.parse(process.env.VARIANTS);
const wait = (ms) => new Promise((r) => setTimeout(r, ms));
const shots = [
  { key: 'night', hour: 21, p: [0, 1.175, 0.16], t: [-0.05, 1.03, -0.95], fov: 54 },
  { key: 'dusk', hour: 18.5, p: [0, 1.175, 0.16], t: [-0.05, 1.03, -0.95], fov: 54 },
  { key: 'day', hour: 13, p: [0, 1.175, 0.16], t: [-0.05, 1.03, -0.95], fov: 54 },
  { key: 'cube', hour: 13, p: [-0.18, 0.95, -0.42], t: [-0.33, 0.765, -0.62], fov: 28 },
];
const browser = await chromium.launch({ executablePath: '/Applications/Google Chrome 3.app/Contents/MacOS/Google Chrome', headless: true, args: ['--use-angle=metal', '--ignore-gpu-blocklist'] });
for (const [name, q] of Object.entries(variants)) {
  for (const s of shots) {
    const page = await browser.newPage({ viewport: { width: 1200, height: 750 } });
    await page.goto(`http://127.0.0.1:5173/?debug&pr=1&hour=${s.hour}&${q}`, { waitUntil: 'load' });
    await page.waitForFunction(() => window.__room !== undefined, null, { timeout: 60000 });
    await page.evaluate((v) => window.__room.parkCamera(v.p, v.t, v.fov), s);
    await wait(1600);
    await page.screenshot({ path: `${out}/${s.key}-${name}.png` });
    await page.close();
  }
}
await browser.close();

File: scripts/bloom-score.py
# Companion to bloom-score.mjs. Usage: python3 scripts/bloom-score.py <outDir> <reference> <nobloom> names...
import sys, colorsys
from PIL import Image, ImageChops
out, ref, nob, *names = sys.argv[1:]
cube_ref = Image.open(f'{out}/cube-{nob}.png').convert('RGB')
w, h = cube_ref.size
px = cube_ref.load()
mask = [(x, y) for y in range(0, h, 2) for x in range(0, w, 2)
        if (lambda hls: hls[2] > 0.45 and 0.12 < hls[1] < 0.85)(colorsys.rgb_to_hls(*[c / 255 for c in px[x, y]]))]
def sat(img):
    p = img.load(); s = 0
    for x, y in mask: s += colorsys.rgb_to_hls(*[c / 255 for c in p[x, y]])[2]
    return s / len(mask)
def diff(a, b):
    d = ImageChops.difference(a.convert('L'), b.convert('L')).crop((0, 0, a.width, a.height - 70))
    v = list(d.get_flattened_data()); return sum(v) / len(v)
print(f"{'variant':34s} cube-sat  night-diff  dusk-diff  day-diff")
for n in [ref, nob] + names:
    c = sat(Image.open(f'{out}/cube-{n}.png').convert('RGB'))
    nd = diff(Image.open(f'{out}/night-{n}.png'), Image.open(f'{out}/night-{ref}.png'))
    dd = diff(Image.open(f'{out}/dusk-{n}.png'), Image.open(f'{out}/dusk-{ref}.png'))
    da = diff(Image.open(f'{out}/day-{n}.png'), Image.open(f'{out}/day-{ref}.png'))
    print(f'{n:34s} {c:.3f}     {nd:6.2f}      {dd:6.2f}     {da:6.2f}')
